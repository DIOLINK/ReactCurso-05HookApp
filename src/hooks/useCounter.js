import { useState } from 'react';

const useCounter = (initialState = 10) => {
  const [counter, setCounter] = useState(initialState);
  const increment = () => setCounter(counter + 1);
  const decrement = () => setCounter(counter - 1);
  const reset = () => setCounter(initialState);

  return { increment, decrement, reset, counter };
};

export default useCounter;
