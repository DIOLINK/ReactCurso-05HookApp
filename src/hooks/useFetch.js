import { useEffect, useRef, useState } from 'react';

const useFetch = (url) => {
  const isMounted = useRef(true);
  const [state, setState] = useState({
    data: null,
    loading: true,
    error: null,
  });

  useEffect(() => {
    return () => {
      isMounted.current = false;
    };
  }, []);
  useEffect(() => {
    setState({
      data: null,
      loading: true,
      error: null,
    });

    if (isMounted.current) {
      fetch(url)
        .then((res) => res.json())
        .then((data) => {
          setState({ data, loading: false, error: null });
        });
    }
  }, [url]);
  return state;
};

export default useFetch;
