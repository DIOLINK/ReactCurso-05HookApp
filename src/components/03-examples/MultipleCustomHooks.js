import React from 'react';
import useCounter from '../../hooks/useCounter';
import useFetch from '../../hooks/useFetch';
import '../02-useEffect/effects.css';
const MultipleCustomHooks = () => {
  const { counter, increment } = useCounter(1);
  const url = `https://www.breakingbadapi.com/api/quotes/${counter}`;
  const { loading, data } = useFetch(url);
  const { author, quote } = !!data && data[0];

  return (
    <div>
      <h1>BreakingBad Quotes!</h1>
      <hr />
      {loading ? (
        <div className="alert alert-info text-center">Loading...</div>
      ) : (
        <figure className="text-end">
          <blockquote className="blockquote">
            <p className="mb-3">{quote}</p>
            <figcaption className="blockquote-footer">{author}</figcaption>
          </blockquote>
        </figure>
      )}
      <button className="btn btn-primary" onClick={increment}>
        Next quote
      </button>
    </div>
  );
};

export default MultipleCustomHooks;
