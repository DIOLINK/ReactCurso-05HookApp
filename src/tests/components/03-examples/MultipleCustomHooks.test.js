import { shallow } from 'enzyme';
import MultipleCustomHooks from '../../../components/03-examples/MultipleCustomHooks';
import useCounter from '../../../hooks/useCounter';
import useFetch from '../../../hooks/useFetch';
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');
describe('Pruebas en <MultipleCustomHooks />', () => {
  it('Debe mostrar el coponente', () => {
    useCounter.mockReturnValue({ counter: 1, increment: () => {} });
    useFetch.mockReturnValue({ data: null, loading: true, error: null });
    const wrapper = shallow(<MultipleCustomHooks />);
    expect(wrapper).toMatchSnapshot();
  });

  it('Debe de mostrar la informacion', () => {
    useCounter.mockReturnValue({ counter: 1, increment: () => {} });
    useFetch.mockReturnValue({
      data: [{ author: 'Roger', quote: 'Te felicito' }],
      loading: false,
      error: null,
    });
    const wrapper = shallow(<MultipleCustomHooks />);
    // console.log('wrapper :>> ', wrapper.html());
    expect(wrapper.find('.alert').exists()).toBe(false);
    expect(wrapper.find('figcaption').text().trim()).toBe('Roger');
  });
});
