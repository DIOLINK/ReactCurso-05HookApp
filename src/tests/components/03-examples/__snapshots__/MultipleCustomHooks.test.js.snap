// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks /> Debe mostrar el coponente 1`] = `
<div>
  <h1>
    BreakingBad Quotes!
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Next quote
  </button>
</div>
`;
